### PR TITLE
improve: consolidate user and machine identities into a single entity

### LINKED ISSUE #1657: Consolidate user and machine identities into a single entity
User and machine identities are similar in many ways but differ in a few key areas, namely authentication.

Users:
- Requires an email (`local@domain`)
- Email + password
- No access key

Machines:
- Requires a name (`^[a-zA-Z0-9-_./]+$`)
- Access key
- No email + password

### Proposal

Both users and machines already exist as a single abstract resource called identity. Instead, identity should be the sole resource representing an identity without user or machine subtypes.

- Identifier can be either an email (`local@domain`) or a name (`^[a-zA-Z0-9-_./]+$`)
- May or may not have a password. No password set disables login with email + password
- Can create and use access keys

### PR DESCRIPTION
**Pull request by BruceMacD**
## Summary
Make identity the sole resource representing an identity without user or machine subtypes.
- Identifier can be either an email (local@domain) or a name (^[a-zA-Z0-9-_./]+$)
- May or may not have a password. No password set disables login with email + password
- Can create and use access keys

## Checklist

- [x] Wrote appropriate unit tests
- [x] Considered security implications of the change
- [x] Updated associated docs where necessary
- [x] Updated associated configuration where necessary
- [x] Change is backwards compatible if it needs to be (user can upgrade without manual steps?)
- [x] Nothing sensitive logged
- [x] Considered data migrations for smooth upgrades

## Related Issues

Resolves #1657

### CODE REVIEW

File: internal/server/server_test.go [hunk 1]
```diff
@@ -371,7 +371,7 @@ func TestServer_PersistSignupUser(t *testing.T) {
 	passwd := "supersecretpassword"
 
 	// run signup for "admin@email.com"
-	signupReq := api.SignupRequest{Email: email, Password: passwd}
+	signupReq := api.SignupRequest{Name: email, Password: passwd}
```
Comment (jmorganca) [inline internal/server/server_test.go:374]: Is this a breaking change for CLI <> API?

Comment (jmorganca) [reply internal/server/server_test.go:374]: nice!

**Review by jmorganca — APPROVED** [review]
Great cleanup

File: Makefile [hunk 2]
```diff
@@ -10,6 +10,10 @@ test:
 test-all:
 	go test ./...
 
+# update the expected command output file
```
[the remark this answers is not in the document]
Comment (dnephin) [reply Makefile:13]: I'm added support for using just the `-update` flag, but haven't tagged a release yet. Hopefully that will make it much easier to remember.

Also doesn't hurt to have it in the Makefile.

File: internal/server/config.go [hunk 3]
```diff
@@ -678,6 +679,7 @@ func (Server) loadGrant(db *gorm.DB, input Grant) (*models.Grant, error) {
 
 		id = uid.NewGroupPolymorphicID(group.ID)
 
+	// TODO: this gotta go
```
Comment (dnephin) [inline internal/server/config.go:682]: Is this TODO still relevant?

File: internal/server/handlers_test.go [hunk 4]
```diff
@@ -307,24 +307,42 @@ func TestAPI_DeleteProvider(t *testing.T) {
...
 	opts.Grants = append(opts.Grants, Grant{
-		Machine:  "admin",
+		User:     "admin",
 		Role:     "admin",
 		Resource: "infra",
 	})
 }
 
-func TestAPI_CreateIdentity(t *testing.T) {
+func TestDeleteProvider_NoDeleteInternalProvider(t *testing.T) {
```
Comment (dnephin) [inline internal/server/handlers_test.go:323]: I think this is covered by the `"infra provider can not be deleted"` case in the table test above, no?

File: internal/server/errors_test.go [hunk 5]
```diff
@@ -60,36 +60,6 @@ func TestSendAPIError(t *testing.T) {
...
-				Code:    http.StatusBadRequest,
-				Message: `Email: failed the "email" check`,
-				FieldErrors: []api.FieldError{
-					{
-						FieldName: "Email",
-						Errors:    []string{`failed the "email" check`},
-					},
-				},
-			},
-		},
```
Comment (dnephin) [inline internal/server/errors_test.go:92]: These test cases a are still relevant, they were testing that validation errors return the right HTTP response code. I think we still want that, we just need a different field to use.

Comment (dnephin) [reply internal/server/errors_test.go:92]: I guess we don't need both, but at least one would be good.

**Pull request merged**